packages/dds/tree/src/core/undo/undoRedoManager.ts

```diff
@@ -0,0 +1,97 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ChangeFamily, ChangeFamilyEditor } from "../change-family";
+import { GraphCommit, GraphCommitType, tagChange } from "../rebase";
+import { ReadonlyRepairDataStore, RepairDataStore } from "../repair";
+
+/**
+ * Manages the undo commit tree and repair data associated with undoable commits.
+ */
+export class UndoRedoManager<TChange, TEditor extends ChangeFamilyEditor> {
+	/**
+	 * @param repairDataStoryFactory - Factory function for creating {@link RepairDataStore}s to create and store repair
+	 * data for {@link UndoableCommit}s.
+	 * @param changeFamily - {@link ChangeFamily} used for inverting changes.
+	 * @param headUndoableCommit - Optional commit to set as the initial undoable commit.
+	 */
+	public constructor(
+		private readonly repairDataStoryFactory: () => RepairDataStore,
+		private readonly changeFamily: ChangeFamily<TEditor, TChange>,
+		protected headUndoableCommit?: UndoableCommit<TChange>,
+	) {}
+
+	/**
+	 * Adds the provided commit to the undo commit tree.
+	 * Should be called for all commits on the relevant branch, including undo commits.
+	 */
+	public trackCommit(commit: GraphCommit<TChange>, repairDataStore?: RepairDataStore): void {
+		if (commit.type === GraphCommitType.Undo) {
+			// Currently no need to handle undo commits
+			return;
+		}
+
+		const parent = this.headUndoableCommit;
+		// If repair data was not provided, create a new repair data store and capture the repair data for the commit
+		const repairData = repairDataStore ?? this.repairDataStoryFactory();
+		if (repairDataStore === undefined) {
+			repairData.capture(this.changeFamily.intoDelta(commit.change), commit.revision);
+		}
+
+		this.headUndoableCommit = {
+			commit,
+			parent,
+			repairData,
+		};
+	}
+
+	/**
+	 * Inverts the head undo commit and applies it as a local change.
+	 */
+	public undo(): TChange | undefined {
+		const commitToUndo = this.headUndoableCommit;
+
+		if (commitToUndo === undefined) {
+			// No undoable commits, exit early
+			return undefined;
+		}
+
+		const { commit, parent, repairData } = commitToUndo;
+		// Removes this undo from the undo commit tree
+		this.headUndoableCommit = parent;
+
+		return this.changeFamily.rebaser.invert(
+			tagChange(commit.change, commit.revision),
+			false,
+			repairData,
+		);
+	}
```

Comment:
The doc comment says the method applies the undo, but that's not quite right: we return a changeset that the call may or may not apply. What we do however is update `this.headUndoableCommit` based on the assumption that the returned changeset will be applied.

I think shifting the contract for this method so that it doesn't make such assumption would make the class easier to understand and the system as a whole less bug-prone.


```suggestion
	/**
	 * @returns a changeset that, if applied, would undo the last undoable commit.
	 * Returns undefined if there is no such commit.
	 */
	public undo(): TChange | undefined {
		const commitToUndo = this.headUndoableCommit;

		if (commitToUndo === undefined) {
			// No undoable commits, exit early
			return undefined;
		}

		const { commit, repairData } = commitToUndo;

		return this.changeFamily.rebaser.invert(
			tagChange(commit.change, commit.revision),
			false,
			repairData,
		);
	}
```

This will require updating the `headUndoableCommit` pointer when `trackCommit` is called with a `GraphCommitType.Undo` commit.